### PR TITLE
Add evaluation configs under qwen2 dir

### PR DESCRIPTION
**Pull request by joecummings**
#### Context
What is the purpose of this PR? Is it to
- [ ] add a new feature
- [ ] fix a bug
- [ ] update tests and/or documentation
- [x] clean up

This PR addresses the changes detailed in #

#### Changelog
* Copied `evaluation.yaml` to `qwen2/` directory
* Updated `evaluation.yaml` to point to Qwen2 7B model instantiations
* Updated the recipe registry to pick up the new config

#### Test plan
- [x] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [ ] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [ ] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [x] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [x] manually run any new or modified recipes with sufficient proof of correctness
- [x] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

**Qwen2 Eleuther eval recipe output**:
```
(joe-torchtune) [jrcummings@devvm050.nha0 ~/projects/joe-torchtune (move-qwen2-eval-and-generate)]$ tune run eleuther_eval --config qwen2/evaluation
2024-10-11:05:39:28,144 INFO     [_logging.py:101] Running EleutherEvalRecipe with resolved config:

batch_size: 8
checkpointer:
  _component_: torchtune.training.FullModelHFCheckpointer
  checkpoint_dir: /tmp/Qwen2-7B-Instruct
  checkpoint_files:
  - model-00001-of-00004.safetensors
  - model-00002-of-00004.safetensors
  - model-00003-of-00004.safetensors
  - model-00004-of-00004.safetensors
  model_type: QWEN2
  output_dir: /tmp/Qwen2-7B-Instruct-finetune
device: cuda
dtype: bf16
enable_kv_cache: true
limit: null
max_seq_length: 4096
model:
  _component_: torchtune.models.qwen2.qwen2_7b
quantizer: null
seed: 1234
tasks:
- truthfulqa_mc2
tokenizer:
  _component_: torchtune.models.qwen2.qwen2_tokenizer
  max_seq_len: null
  merges_file: /tmp/Qwen2-7B-Instruct/merges.txt
  path: /tmp/Qwen2-7B-Instruct/vocab.json

Model is initialized with precision torch.bfloat16.
2024-10-11:05:39:35,359 INFO     [eleuther_eval.py:524] Model is initialized with precision torch.bfloat16.
2024-10-11:05:39:35,821 INFO     [huggingface.py:129] Using device 'cuda:0'
2024-10-11:05:39:36,009 INFO     [huggingface.py:481] Using model type 'default'
2024-10-11:05:39:36,297 INFO     [huggingface.py:365] Model parallel was set to False, max memory was not set, and device map was set to {'': 'cuda:0'}
Running evaluation on the following tasks: ['truthfulqa_mc2']
2024-10-11:05:39:49,460 INFO     [eleuther_eval.py:568] Running evaluation on the following tasks: ['truthfulqa_mc2']
2024-10-11:05:39:49,462 INFO     [task.py:415] Building contexts for truthfulqa_mc2 on rank 0...
2024-10-11:05:39:50,779 INFO     [evaluator.py:489] Running loglikelihood requests
Running loglikelihood requests: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 5882/5882 [01:32<00:00, 63.67it/s]
2024-10-11:05:41:28,550 INFO     [eleuther_eval.py:577] Eval completed in 99.09 seconds.
2024-10-11:05:41:28,550 INFO     [eleuther_eval.py:578] Max memory allocated: 42.86 GB


|    Tasks     |Version|Filter|n-shot|Metric|   |Value |   |Stderr|
|--------------|------:|------|-----:|------|---|-----:|---|-----:|
|truthfulqa_mc2|      2|none  |     0|acc   |↑  |0.5729|±  |0.0154|
```

### CODE REVIEW

File: recipes/configs/qwen2/evaluation.yaml [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+checkpointer:
+  _component_: torchtune.training.FullModelHFCheckpointer
+  checkpoint_dir: /tmp/Qwen2-7B-Instruct
+  checkpoint_files: [
+    model-00001-of-00004.safetensors,
+    model-00002-of-00004.safetensors,
+    model-00003-of-00004.safetensors,
+    model-00004-of-00004.safetensors
+  ]
+  output_dir: ./ # Not needed
```
Comment (ebsmothers) [inline recipes/configs/qwen2/evaluation.yaml:20]: 👀

**Review by ebsmothers — COMMENTED** [review]
What's up with the <|im_end|> token in the generation outputs?

File: recipes/configs/qwen2/evaluation.yaml (hunk 1 above)
Comment (joecummings) [reply recipes/configs/qwen2/evaluation.yaml:20]: There's nothing to OUTPUT.

**Comment by joecummings** [comment]
> What's up with the <|im_end|> token in the generation outputs?

I can look into it, but this isn't part of the change. It was here before and after the change. Also, this is the *old* generation recipe - we aren't tackling that now. So results are not expected to be *great*.

Stamp please.

**Comment by SalmanMohammadi** [comment]
Wait, why not just add a config for generate V2 instead?

**Comment by joecummings** [comment]
> Wait, why not just add a config for generate V2 instead?

I wanted to do things in stages so (1) move everything while at the same time (1b) adding compile support and quantization to generate_v2 then (2) switch to generate_v2.

I could be being pedantic though.

File: recipes/configs/qwen2/evaluation.yaml (hunk 1 above)
Comment (ebsmothers) [reply recipes/configs/qwen2/evaluation.yaml:20]: Yeah I just mean this is an indication of a gap in the current checkpointer, this should clearly not be a mandatory arg

**Comment by ebsmothers** [comment]
Yeah you might inevitably be making more work for yourself (or whomever else) down the line, as there will now be N configs to migrate instead of 1. Either way, I'm OK with it. At least this way we explicitly have configs for all the different models before the migration and so are more likely to find any gaps earlier

**Review by ebsmothers — APPROVED** [review]

**Comment by SalmanMohammadi** [comment]
I also just saw the parent issue. I love the momentum you're generating here and I wish to nudge it in a direction that means we arrive at deprecating the old generation recipe (yuck) sooner, and whilst minimizing maintenance and testing overhead.

**Comment by SalmanMohammadi** [comment]
> > I also just saw the parent issue. I love the momentum you're generating here and I wish to nudge it in a direction that means we arrive at deprecating the old generation recipe (yuck) sooner, and whilst minimizing maintenance and testing overhead.
> 
> Are you OK migrating to generate_v2 without compile support? If yes, I'll change the parent issue and this PR to directly migrate to V2. If not, then this is the stopgap while we get that working.

Personally, yes, I don't think compile support is critical for a single-batch-element, one-time inference script. I think quantized model inference is the more crtitical gap to address, but it does work. That's just my 2 cents tho.

I'd volunteer to address gaps in time for when we deprecate if needed.

**Review by SalmanMohammadi — APPROVED** [review]

**Comment by joecummings** [comment]
Okay based on the comments, here's my plan:

1. Update the parent issue to *only* moving the eleuther eval scripts
2. Update this PR accordingly
3. Open a separate tracker to migrate all models off the generate.py script to the generate_v2.py script

In the meantime, @SalmanMohammadi  and I will work on adding compile and quantization support for generate_v2.

Good?

**Comment by SalmanMohammadi** [comment]
image
